docs/extensions/readme.md

```diff
@@ -1173,3 +1175,56 @@ An object with an optional property.
   }
 }
 ```
+
+## x-ms-header-collection-prefix
+Use `x-ms-header-collection-prefix` to deal with prefixes for headers. Has different behavior if it refers to a request header or a response header:
+
+- Request header: All keys in the request headers will be prefixed with the prefix value before being sent to the service.
+- Response header: Only response headers that start with the prefix specified here will be returned to users. Additionally, the prefix will be stripped from the response header key before being returned to users
+
+Additionally, applying this extension to a schema forces the schema to become a dictionary.
+
+**Schema**: `string`. Name of the prefix you want to append / filter by. A common value for storage libraries is `x-ms-meta-`.
+
+**Request Example**:
+
+```json
+"parameters": [
+  "x-ms-meta": {
+    "type": "string",
+    "x-ms-client-name": "Metadata",
+    "x-ms-header-collection-prefix": "x-ms-meta-"
+  }
+]
+```
+This request parameter will be forced to be a dictionary schema, and all keys in this dictionary will be prefixed with `x-ms-meta-`.
+So, if you input `{"key": "value"}` through this parameter, `{"x-ms-meta-key": "value"}` is what reaches the service.
+
+**Response Example**:
+
+```json
+"responses": {
+  "200": {
+    "description": "Success",
+    "headers": {
+      "x-ms-meta": {
+        "type": "string",
+        "x-ms-client-name": "Metadata",
+        "x-ms-header-collection-prefix": "x-ms-meta-"
+      }
+    }
+  }
+}
+```
+
+This response header parameter will be forced to be a dictionary schema. Only entries with prefix `x-ms-meta-` will be returned to users,
```

Comment:
Just so I completely understand. 
Let say I have a header `foo` defined as a string. Then in the generated it should provide a way to give a string value for that header as expected.
Now however if I add the `x-ms-header-collection-prefix` property it will automatically change the type of that header from string to `Dict<string, string>`?

Reply:
Correct

Reply:
basically if you have a header parameter, even if it's schema is defined to be of type string, m4 should model it as a string dictioniary instead

this is based off the issue @pakrym opened for this https://github.com/Azure/autorest/issues/3761

Reply:
yeah so regardless of the documentation, if modeler4 decide to change that automatically is that going to create a breaking change for language already doing it?
would it convert `string` (m4)-> `Dict<string, string>` (lang)-> `Dict<string, Dict<string, string>>`

Reply:
We would need to make sure everyone is aware of the change and ready to react. I suspect C# and GO are the only two languages that would need to.

Reply:
also @timotheeguerin , pthon is the only language generator that has GA'd, and we don't have support foer it yet, so not an issue in terms for breaking GA'd libraries

Reply:
If I'm following your question correctly - the purpose of this extension is that it's for undefined headers. The SDK/swagger has no knowledge of what the header names might be.
So yes - it would be a collection/dictionary of arbitrary data, that was gathered from headers marked with the requisite prefix.